src/main.go

```diff
@@ -25,90 +27,144 @@ type SubjectAccessReviewAPI struct {
+	return func(w http.ResponseWriter, r *http.Request) {
 
-		// Print dump for logging
 		dump, dumperr := httputil.DumpRequest(r, true)
 		if dumperr != nil {
 			fmt.Println("Error dumping request:", dumperr)
 			return
 		}
 
-		if(logLevel >= 2) { fmt.Println(string(dump)) }
-
 		var sar SubjectAccessReviewAPI
 		err := json.NewDecoder(r.Body).Decode(&sar)
 		if err != nil {
-			fmt.Println("[ERROR] ",err.Error())
-			http.Error(w, "Invalid JSON", http.StatusBadRequest)
+			fmt.Println("JSON decoding error: ", err.Error())
+			http.Error(w, "JSON decoding error: "+err.Error(), http.StatusBadRequest)
 			return
 		}
 
 		defer r.Body.Close()
 
-		isUnprivilegedUser := slices.Contains(sar.Spec.Groups, unprivilegedGroup) || slices.Contains(sar.Spec.Group, unprivilegedGroup)
-		isProtectedNamespace := sar.Spec.ResourceAttributes != nil && slices.Contains(protectedNamespaces, sar.Spec.ResourceAttributes.Namespace) // TODO: test if you can bypass with empty or all namespaces
-		isSecret := sar.Spec.ResourceAttributes != nil && sar.Spec.ResourceAttributes.Resource == "secrets"                                       //TODO: test if you can bypass with * or singular nouns
-		isReadonlyVerb := sar.Spec.ResourceAttributes != nil && slices.Contains(readonlyVerbs, sar.Spec.ResourceAttributes.Verb)
+		authorized, denyReason := isRequestAuthorized(sar, protectedNamespaces, additionalPrivilegedUsers)
 
 		status := new(authorizationv1.SubjectAccessReviewStatus)
-		status.Allowed = false
-		if isUnprivilegedUser && isProtectedNamespace && isSecret {
-			status.Denied = true
-			status.Reason = "Cannot access secrets in protected namespace"
-		} else if isUnprivilegedUser && isProtectedNamespace && !isReadonlyVerb {
-			status.Denied = true
-			status.Reason = "Cannot write to protected namespace"
-		} else {
-			status.Denied = false
-			status.Allowed = opinionMode
-			if(!opinionMode){ status.Reason = "Webhook doesn't give opinion, delegated to other authorizers" }
+		status.Denied = !authorized
+		status.Allowed = opinionMode && authorized
+
+		if status.Denied {
+			status.Reason = denyReason
+		} else if !opinionMode {
+			status.Reason = "Webhook doesn't give opinion, delegated to other authorizers"
 		}
-		//todo: add status.EvaluationError handling
 
 		responseReview := new(SubjectAccessReviewHTTPResponse)
 		responseReview.ApiVersion = "authorization.k8s.io/v1"
 		responseReview.Kind = "SubjectAccessReview"
 		responseReview.Status = *status
 
-		if(status.Denied && logLevel == 1) { fmt.Println(string(dump)) }
-		if(status.Denied && logLevel >= 1){ fmt.Printf("[DENIED] Reason: %s\n",status.Reason) }
+		var deniedLogOutput string
+		if status.Denied {
+			deniedLogOutput = "Denied"
+		} else {
+			deniedLogOutput = "Allowed"
+		}
+		if logLevel >= 1 && sar.Spec.NonResourceAttributes != nil {
+			fmt.Printf("%s non-resource request from \"%s\". Reason: %s\n", deniedLogOutput, sar.Spec.User, status.Reason)
```

Comment:
Lets move this to the golang Log system, and try to use Println() to make sure they all go onto a new line:
https://pkg.go.dev/log

Its worth checking that we get what we expect to see inside Loki.

Reply:
I think the request has RemoteAddr and the Headers, which should tell us enough so we know which cluster has sent the request. Needs testing to see what works best inside Loki.

Key take away: be clear which cluster sent the request.